File: zZ/cpp/longest-increasing-path-in-a-matrix.cpp
// 329. Longest Increasing Path in a Matrix
// Difficulty: Hard

// Given an integer matrix, find the length of the longest increasing path.

// From each cell, you can either move to four directions: left, right, up or down.
// You may NOT move diagonally or move outside of the boundary (i.e. wrap-around is not allowed).

// Example 1:

// nums = [
//   [9,9,4],
//   [6,6,8],
//   [2,1,1]
// ]
// Return 4
// The longest increasing path is [1, 2, 6, 9].

// Example 2:

// nums = [
//   [3,4,5],
//   [3,2,6],
//   [2,2,1]
// ]
// Return 4
// The longest increasing path is [3, 4, 5, 6]. Moving diagonally is not allowed.

// Time:  O(m * n)
// Space: O(m * n)

// DFS + Memorization solution.
class Solution {
public:
  int longestIncreasingPath(vector<vector<int>>& matrix) {
    int result = 0;
    vector<vector<int>> maxLengths(matrix.size(), vector<int>(matrix[0].size()));
    for (int i = 0; i < matrix.size(); ++i) {
      for (int j = 0; j < matrix[0].size(); ++j) {
        result = max(result, longestpath(matrix, i, j, maxLengths));
      }
    }
    return result;
  }

private:
  int longestpath(const vector<vector<int>>& matrix, const int i, const int j, vector<vector<int>>& maxLengths) {
    if (maxLengths[i][j]) {
      return maxLengths[i][j];
    }    
    int maxDepth = 1;
    for (auto pos : vector<pair<int, int>> {{i, j - 1}, {i, j + 1}, {i - 1, j}, {i + 1, j}}) {
      const int x = pos.first, y = pos.second;
      if (x >= 0 && x < matrix.size() && y >= 0 && y < matrix[0].size() && matrix[x][y] > matrix[i][j]) {
        maxDepth = max(maxDepth, longestpath(matrix, x, y, maxLengths) + 1);
      }
    }
    return maxLengths[i][j] = maxDepth;
  }
};

File: data-structure-and-algorithm/typescript/leetcode/src/longest-increasing-path-in-a-matrix.ts
/*
Given an m x n integers matrix, return the length of the longest increasing path in matrix.

From each cell, you can either move in four directions: left, right, up, or down. You may not move diagonally or move outside the boundary (i.e., wrap-around is not allowed).



Example 1:


Input: matrix = [[9,9,4],[6,6,8],[2,1,1]]
Output: 4
Explanation: The longest increasing path is [1, 2, 6, 9].
Example 2:


Input: matrix = [[3,4,5],[3,2,6],[2,2,1]]
Output: 4
Explanation: The longest increasing path is [3, 4, 5, 6]. Moving diagonally is not allowed.
Example 3:

Input: matrix = [[1]]
Output: 1


Constraints:

m == matrix.length
n == matrix[i].length
1 <= m, n <= 200
0 <= matrix[i][j] <= 231 - 1
*/

function longestIncreasingPath(matrix) {
  const [rows, cols] = [matrix.length, matrix[0].length];
  const inDegree = [...Array(rows)].map(() => Array(cols).fill(0));
  for (let i = 0; i < rows; ++i) {
    for (let j = 0; j < cols; ++j) {
      for (const [dx, dy] of [
        [-1, 0],
        [1, 0],
        [0, 1],
        [0, -1],
      ]) {
        const [x, y] = [i + dx, j + dy];
        if (x >= 0 && x < rows && y >= 0 && y < cols) {
          if (matrix[x][y] > matrix[i][j]) {
            ++inDegree[x][y];
          }
        }
      }
    }
  }

  console.log(inDegree);

  let q = [];
  for (let i = 0; i < rows; ++i) {
    for (let j = 0; j < cols; ++j) {
      if (!inDegree[i][j]) {
        q.push([i, j]);
      }
    }
  }

  let result = 0;
  for (; q.length; ++result) {
    const next = [];
    for (const [i, j] of q) {
      for (const [dx, dy] of [
        [-1, 0],
        [1, 0],
        [0, 1],
        [0, -1],
      ]) {
        const [x, y] = [i + dx, j + dy];
        if (x >= 0 && x < rows && y >= 0 && y < cols) {
          if (matrix[x][y] > matrix[i][j]) {
            if (--inDegree[x][y] === 0) {
              next.push([x, y]);
            }
          }
        }
      }
    }
    q = next;
  }
  return result;
}

function longestIncreasingPathDFS(matrix) {
  const cache = [...Array(matrix.length)]
    .map(() => Array(matrix[0].length))
    .fill(0);
  let result = 0;
  for (let i = 0; i < matrix.length; ++i) {
    for (let j = 0; j < n; ++j) {
      const length = dfs(matrix, i, j, cache);
      result = Math.max(result, length);
    }
  }
  return result;
}

function dfs(matrix, i, j, cache) {
  if (cache[i][j]) {
    return cache[i][j];
  }
  let max = 1;
  for (const [dx, dy] of [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ]) {
    const [x, y] = [i + dx, j + dy];
    if (
      x >= 0 &&
      x < matrix.length &&
      y >= 0 &&
      y < matrix[0].length &&
      matrix[x][y] > matrix[i][j]
    ) {
      const length = 1 + dfs(matrix, x, y, cache);
      max = Math.max(max, length);
    }
  }
  cache[i][j] = max;
  return max;
}
